core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -537,23 +470,130 @@ public long count() {
+      Thread thread = Thread.currentThread();
+      executor.execute(() -> {
+         try {
+            log.tracef("Thread %s submitted iterator request for stream", thread);
+            Set<Integer> segmentsToProcess = segmentsToFilter == null ?
+                    new ReplicatedConsistentHash.RangeSet(segmentInfoCH.getNumSegments()) : segmentsToFilter;
+            do {
+               ConsistentHash ch = dm.getReadConsistentHash();
+               boolean runLocal = ch.getMembers().contains(localAddress);
+               Set<Integer> segments;
+               Set<Object> excludedKeys;
+               boolean stayLocal = false;
+               if (runLocal) {
+                  Set<Integer> segmentsForOwner = ch.getSegmentsForOwner(localAddress);
+                  stayLocal = segmentsToFilter != null && segmentsForOwner.containsAll(segmentsToFilter);
+                  segments = stayLocal ? segmentsForOwner : ch.getPrimarySegmentsForOwner(localAddress);
+                  segments.retainAll(segmentsToProcess);
+
+                  excludedKeys = segments.stream().flatMap(s -> results.referenceArray.get(s).stream())
+                          .collect(Collectors.toSet());
+               } else {
+                  segments = null;
+                  excludedKeys = Collections.emptySet();
+               }
+               KeyTrackingTerminalOperation<Object, R, Object> op = iteratorOperation.getOperation(
+                       intermediateOperations, supplierForSegments(ch, segmentsToProcess, excludedKeys, !stayLocal),
+                       distributedBatchSize);
+               if (!stayLocal) {
+                  UUID id = csm.remoteStreamOperationRehashAware(iteratorParallelDistribute, parallel, ch,
+                          segmentsToProcess, keysToFilter, new AtomicReferenceArrayToMap<>(results.referenceArray),
+                          includeLoader, op, results);
+                  supplier.pending = id;
+                  try {
+                     if (runLocal) {
+                        performLocalRehashAwareOperation(results, segmentsToProcess, ch, segments, op,
+                                () -> ch.getPrimarySegmentsForOwner(localAddress), id);
+                     }
+                     try {
+                        if (!csm.awaitCompletion(id, 30, TimeUnit.SECONDS)) {
+                           throw new TimeoutException();
+                        }
+                     } catch (InterruptedException e) {
+                        throw new CacheException(e);
+                     }
+                     segmentsToProcess = segmentsToProcess(supplier, results, segmentsToProcess, id);
+                  } finally {
+                     csm.forgetOperation(id);
+                  }
+               } else {
+                  performLocalRehashAwareOperation(results, segmentsToProcess, ch, segments, op,
+                          () -> ch.getSegmentsForOwner(localAddress), null);
```

Comment:
We can just return the segmentsForOwner from https://github.com/infinispan/infinispan/pull/3614/files#diff-be6f435a8edfd997f62d9a90ad59582bR514 if you push the variable out of the if brackets.  It can be defined as null and then only set if runLocal was set since we know stayLocal can only be true if it went into runLocal.  This will avoid an additional segmentsForOwner lookup.